### PR TITLE
feat: add snowflake sync workflow

### PR DESCRIPTION
**Pull request by fontanierh**
## Description

Adds the logic for the snowflake sync activity. Add also the missing `upsertTable` util for connectors (we are relying on `upsertTableFromCsv` for notion / ghseet / msft)

## Risk

N/A behind feature flag for now

## Deploy Plan

Deploy connectors and front

### CODE REVIEW

File: connectors/src/connectors/snowflake/temporal/activities.ts [hunk 1]
```diff
@@ -1,7 +1,155 @@
...
+          dataSourceConfig: dataSourceConfigFromConnector(connector),
+          tableId: internalId,
+          tableName: internalId,
+          remoteDatabaseTableId: internalId,
+          remoteDatabaseSecretId: connector.connectionId,
+          // TODO(SNOWFLAKE): decide what to do wrt description.
+          tableDescription: "",
+          parents: [
+            `${table.databaseName}.${table.schemaName}`,
+            table.databaseName,
```
Comment (Fraggle) [inline connectors/src/connectors/snowflake/temporal/activities.ts:131]: (does order matter?)

Comment (fontanierh) [reply connectors/src/connectors/snowflake/temporal/activities.ts:131]: I heard that it does

Comment (fontanierh) [reply connectors/src/connectors/snowflake/temporal/activities.ts:131]: But I don't know for sure cc @PopDaph

Comment (Fraggle) [reply connectors/src/connectors/snowflake/temporal/activities.ts:131]: If yes: put a comment.

File: connectors/src/connectors/snowflake/temporal/client.ts [hunk 2]
```diff
@@ -40,11 +40,13 @@ export async function launchSnowflakeSyncWorkflow(
...
       },
       signal: resyncSignal,
+      // If we don't pass signalArgs the workflow will not be signaled.
+      signalArgs: [],
       memo: {
         connectorId,
       },
       // Every 6 hours.
-      cronSchedule: "0 0 */6 * * *",
+      cronSchedule: "0 */6 * * *",
```
Comment (Fraggle) [inline connectors/src/connectors/snowflake/temporal/client.ts:49]: As someone who waited for metabase sync and considering the super fast workflow, I would put it much more frequently.

File: connectors/src/connectors/snowflake/temporal/activities.ts [hunk 3]
```diff
@@ -1,7 +1,157 @@
...
+  if (tablesRes.isErr()) {
+    throw tablesRes.error;
+  }
+
+  const tablesValidation = t.array(snowflakeTableCodec).decode(tablesRes.value);
+  if (isLeft(tablesValidation)) {
+    const pathError = reporter.formatValidationErrors(tablesValidation.left);
+    throw new Error(`Invalid tables response: ${pathError}`);
+  }
+  const [allDatabases, allSchemas, allTables] = await Promise.all([
```
Comment (Fraggle) [inline connectors/src/connectors/snowflake/temporal/activities.ts:45]: nit: I would prefix the things coming from snowflake vs the things coming from our DB to make the code more readable.

File: connectors/src/connectors/snowflake/temporal/activities.ts [hunk 4]
```diff
@@ -1,7 +1,157 @@
...
+  for (const internalId of internalIdsOnSnowflake) {
+    if (isTableReadGranted(internalId)) {
+      let table = tableByInternalId[internalId];
+      if (!table || !table.lastUpsertedAt) {
+        if (!table) {
+          const [dbName, schemaName, tableName] =
+            parseTableInternalId(internalId);
+          table = await RemoteTableModel.create({
+            connectorId,
+            internalId,
```
Comment (Fraggle) [inline connectors/src/connectors/snowflake/temporal/activities.ts:113]: nit: I was confused a bit wether internalId refer to the snowflakeInternalid or not (looking at the code, it's the same thing).

File: connectors/src/connectors/snowflake/temporal/activities.ts [hunk 5]
```diff
@@ -1,7 +1,157 @@
...
+      readGrantedInternalIds.has(tableInternalId)
+    );
+  };
+
+  const tablesOnSnowflake = tablesValidation.right;
+  const internalIdsOnSnowflake = new Set(
+    tablesOnSnowflake.map(
+      (t) => `${t.database_name}.${t.schema_name}.${t.name}`
+    )
+  );
```
Comment (Fraggle) [inline connectors/src/connectors/snowflake/temporal/activities.ts:102]: nit: maybe move this just after the retrieval of tablesValidation ?

File: connectors/src/connectors/snowflake/temporal/activities.ts [hunk 6]
```diff
@@ -1,7 +1,157 @@
...
+      }
+    }
+  }
+
+  for (const t of Object.values(tableByInternalId)) {
+    if (
+      !isTableReadGranted(t.internalId) ||
+      !internalIdsOnSnowflake.has(t.internalId)
+    ) {
+      if (t.lastUpsertedAt) {
```
Comment (Fraggle) [inline connectors/src/connectors/snowflake/temporal/activities.ts:146]: why this check ?

Comment (fontanierh) [reply connectors/src/connectors/snowflake/temporal/activities.ts:146]: if this is null then it means we didn't create the table on core yet so nothing to delete

**Review by Fraggle — APPROVED** [review]
LGTM.

File: connectors/src/connectors/snowflake/temporal/activities.ts (hunk 4 above)
Comment (fontanierh) [reply connectors/src/connectors/snowflake/temporal/activities.ts:112]: Yeah `internalId` is a name convention we use in `connectors` -- agree it's a bit confusing

File: connectors/src/connectors/snowflake/temporal/activities.ts (hunk 1 above)
Comment (PopDaph) [reply connectors/src/connectors/snowflake/temporal/activities.ts:130]: I think so cause of https://github.com/dust-tt/dust/pull/6950

**Pull request merged**